pkg/virt-controller/watch/pool/pool.go

```diff
@@ -1670,3 +1672,76 @@ func (c *Controller) handlePoolDeletion(pool *poolv1.VirtualMachinePool, vms []*
 
 	return nil
 }
+
+func isAutohealingEnabled(pool *poolv1.VirtualMachinePool) bool {
+	return pool.Spec.Autohealing != nil
+}
+
+func (c *Controller) autoHealFailingVMs(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) error {
+	vmsToCleanup := filterFailingVMsToStart(vms, pool.Spec.Autohealing)
+
+	return c.scaleIn(pool, vmsToCleanup, len(vmsToCleanup))
+}
+
+func filterFailingVMsToStart(vms []*virtv1.VirtualMachine, autohealing *poolv1.VirtualMachinePoolAutohealingStrategy) []*virtv1.VirtualMachine {
+	filtered := []*virtv1.VirtualMachine{}
+	for _, vm := range vms {
+		if vm.Status.StartFailure != nil && vm.Status.StartFailure.ConsecutiveFailCount >= getFailureToStartThreshold(autohealing) {
+			filtered = append(filtered, vm)
+			continue
+		}
+
+		if shouldAutohealBasedOnStatus(vm, autohealing) {
+			filtered = append(filtered, vm)
+		}
+	}
+
+	return filtered
+}
+
+// shouldAutohealBasedOnStatus checks if a VM's PrintableStatus indicates it should be autohealed
+func shouldAutohealBasedOnStatus(vm *virtv1.VirtualMachine, autohealing *poolv1.VirtualMachinePoolAutohealingStrategy) bool {
+	switch vm.Status.PrintableStatus {
+	case virtv1.VirtualMachineStatusCrashLoopBackOff,
+		virtv1.VirtualMachineStatusUnschedulable,
+		virtv1.VirtualMachineStatusDataVolumeError,
+		virtv1.VirtualMachineStatusPvcNotFound,
+		virtv1.VirtualMachineStatusErrImagePull,
+		virtv1.VirtualMachineStatusImagePullBackOff:
+		return hasVMBeenFailingLongEnough(vm, autohealing)
+	default:
+		return false
+	}
+}
+
+// hasVMBeenFailingLongEnough checks if VM has not been ready for minimum duration
+func hasVMBeenFailingLongEnough(vm *virtv1.VirtualMachine, autohealing *poolv1.VirtualMachinePoolAutohealingStrategy) bool {
+	for _, condition := range vm.Status.Conditions {
+		if condition.Type == virtv1.VirtualMachineReady && condition.Status == k8score.ConditionFalse {
+			failingSince := condition.LastProbeTime.Time
+
+			if time.Since(failingSince) >= getMinFailingToStartDuration(autohealing) {
+				log.Log.Object(vm).Infof("VM %s/%s has been failing to start for %v, adding to list", vm.Namespace, vm.Name, time.Since(failingSince))
+				return true
+			}
+		}
+	}
+
+	return false
+}
```

Comment:
```suggestion
func hasVMBeenFailingLongEnough(vm *virtv1.VirtualMachine, autohealing *poolv1.VirtualMachinePoolAutohealingStrategy) bool {
	condManager := controller.NewVirtualMachineConditionManager()
	if c := condManager.GetCondition(vm, virtv1.VirtualMachineReady); c != nil && c.Status == k8score.ConditionFalse {
		failingSince := c.LastProbeTime.Time
		if time.Since(failingSince) >= getMinFailingToStartDuration(autohealing) {
			log.Log.Object(vm).Infof("VM %s/%s has been failing to start for %v, adding to list", vm.Namespace, vm.Name, time.Since(failingSince))
			return true
		}
	}
	return false
}
```